### PR TITLE
Add BookCards to Homescreen

### LINKED ISSUE #2: Add BookCards to BookList
### Value statement
__________________________________________________________________________________________

As a **user**
I need to **add books**
so that **I can add books by myself.**

### Description (Text, Scribble, Wireframe, Design)
__________________________________________________________________________________________
Untitled-2020-11-15-1926(10)

### Acceptance criteria
__________________________________________________________________________________________

- [x] Title and Author are required, submit button is disabled, if fields are empty.
- [x] Cover and Description are optional.
- [x] By Submit BookCard is added to BookList.
- [x] Modal are closed by Button.

### Tasks
___________________________________________________________________________________________
- [x] Create feature branch
- [x] Create CircleButton component in Styleguidist
- [x] Create RectangleButton component in Styleguidist
- [x] Create AddBookForm component in Styleguidist
- [x] Write component tests
- [x] Add components to app (feature implementation with state)
- [x] Add missing styles

### Complexity
___________________________________________________________________________________________
medium

### PR DESCRIPTION
**Pull request by CinFrei**
Hi there. :) I created a AddBookForm to add BookCards to the Homescreen. The AddBookForm is triggered by a FloatingActionButton and pops up in a Modal. Please review. :)

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,20 +1,64 @@
+import { useEffect, useState } from 'react'
+import { v4 as uuidv4 } from 'uuid'
 import styled from 'styled-components'
```
Comment (macfrei) [inline src/App.js]: `import styled from 'styled-components/macro'`

File: src/App.js [hunk 2]
```diff
@@ -1,20 +1,64 @@
...
+import AddBookForm from './components/AddBookForm'
+import getBooks from './services/getBooks'
+
 import BookList from './components/BookList'
+import BookCard from './components/BookCard'
 
 export default function App() {
+  const [books, setBooks] = useState(loadLocally('books') ?? [])
+
+  const [addBookFormModal, setaddBookFormModal] = useState(false)
```
Comment (macfrei) [inline src/App.js]: A small typo: `setAddBookFormModal`
```suggestion
  const [addBookFormModal, setAddBookFormModal] = useState(false)
```

File: src/App.js [hunk 3]
```diff
@@ -1,20 +1,64 @@
...
+  function closeAddBookFormModal() {
```
Comment (macfrei) [inline src/App.js]: Why don't you use a toggle function instead of two functions?

File: src/App.js [hunk 4]
```diff
@@ -1,20 +1,64 @@
...
+          title={title}
+          author={author}
+          description={description}
+          key={id}
+        />
+      ))}
     </StyledApp>
   )
+
+  function addBook(title, author, description) {
```
Comment (macfrei) [inline src/App.js]: I'd either have all functions before the `return` statement or all after.

File: src/components/AddBookForm.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
+import styled from 'styled-components/macro'
+
+import Button from './RectangleButton'
+import downArrow from '../assets/arrow-down.svg'
```
Comment (macfrei) [inline src/components/AddBookForm.js]: You could import your svgs like this: `import {ReactComponent as ReactLogo} from './logo.svg';`. It'll then render as svg in your code and you don't need the img element.
```suggestion
import { ReactComponent as DownArrow } from '../assets/arrow-down.svg'
```

File: src/components/FloatingActionButton.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import styled from 'styled-components/macro'
+//import PropTypes from 'prop-types'
```
Comment (macfrei) [inline src/components/FloatingActionButton.js]: If you don't use the import here, then delete the comment.

File: src/App.js [hunk 7]
```diff
@@ -1,20 +1,64 @@
...
+  const [addBookFormModal, setaddBookFormModal] = useState(false)
+  function closeAddBookFormModal() {
+    setaddBookFormModal(false)
+  }
+  function openAddBookFormModal() {
+    setaddBookFormModal(true)
+  }
+
+  useEffect(() => {
+    getBooks().then((books) => setBooks(books))
```
Comment (macfrei) [inline src/App.js:17]: Maybe add a `.catch()` so that an error would actually not cause the app to crash.

File: src/App.js [hunk 8]
```diff
@@ -1,20 +1,64 @@
...
       <h1>Book Owls</h1>
       <BookList listName="Bücherregal" />
+      <FloatingActionButton onClick={openAddBookFormModal} />
+      {addBookFormModal && (
+        <AddBookForm
+          onCreateBook={addBook}
+          onClickDownButton={closeAddBookFormModal}
+        />
+      )}
+      {books.map(({ title, author, description, id }, index) => (
```
Comment (macfrei) [inline src/App.js:36]: Is it intentional that the books you add by yourself are not added to the BookList component? I added a book and it wasn't rendered within the booklist, which makes it look a bit different than the other books, for example there is no padding. 

I would suggest to add `books` to your booklist component as props. It is a bit odd to have `BookCard`s rendered at two different places. If you have a question, please let me know.

**Review by macfrei — COMMENTED** [review]
Nicely done! I added a few comments, they are mostly just minor things. The only issue I have is how you render the bookcards which are created via the form. Please check out my comment about this and if you have any questions, get back to me :) 
Really impressed by what you have accomplished so far and your code looks good as well!

File: src/App.js (hunk 1 above)
Comment (CinFrei) [reply src/App.js]: Thx

File: src/App.js [hunk 9]
```diff
@@ -1,20 +1,64 @@
...
+  useEffect(() => {
+    getBooks().then((books) => setBooks(books))
+  }, [])
+
   return (
     <StyledApp>
       <h1>Book Owls</h1>
       <BookList listName="Bücherregal" />
+      <FloatingActionButton onClick={openAddBookFormModal} />
+      {addBookFormModal && (
```
Comment (felix-horn) [inline src/App.js:33]: Really cool how fast you added this implementation after our discussion yesterday!!

File: src/App.js [hunk 10]
```diff
@@ -1,20 +1,64 @@
...
-  font-family: sans-serif;
-  text-align: center;
+  position: relative;
 
   h1 {
+    margin: 0px;
     color: var(--dark-blue);
+    text-align: center;
   }
 `
```
Comment (felix-horn) [inline src/App.js]: You could think about defining your h1 in the GlobalStyle.js

File: src/components/AddBookForm.js [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+
+export default function AddBookForm({ onCreateBook, onClickDownButton }) {
+  function handleSubmit(event) {
+    event.preventDefault()
+    const formElement = event.target
+    const inputTitle = formElement.elements.title
+    const inputAuthor = formElement.elements.author
+    const inputDesqription = formElement.elements.description
+    onCreateBook(inputTitle.value, inputAuthor.value, inputDesqription.value)
```
Comment (felix-horn) [inline src/components/AddBookForm.js]: typo: "Description"

File: src/components/AddBookForm.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    formElement.reset()
+    inputTitle.focus()
+  }
+
+  return (
+    <StyledModal>
+      <StyledButton onClick={onClickDownButton}>
```
Comment (felix-horn) [inline src/components/AddBookForm.js]: Here I would just say `onClick={onButtonClick}` or even shorter `onClick={onClick}`. Of course, this also had to be adjusted in your App.js (`onClick={closeAddBookFormModal}`)

File: src/components/AddBookForm.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  background-color: transparent;
+  justify-self: end;
+
+  img {
+    width: 25px;
+  }
+`
+
+const FormStyled = styled.form`
+  margin: 10px 10px 20px 10px;
```
Comment (felix-horn) [inline src/components/AddBookForm.js]: margin: 10px 10px 20px;

File: src/components/AddBookForm.test.js [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+      'Die 100 besten Currywurst Rezepte',
+      'Horst Host',
+      'Voll Currywurst. Subba Legga!'
+    )
+
+    expect(getByLabelText('Titel:')).toHaveValue('')
+    expect(getByLabelText('Autor:')).toHaveValue('')
+    expect(getByLabelText('Beschreibung:')).toHaveValue('')
+  })
+  //...
```
Comment (felix-horn) [inline src/components/AddBookForm.test.js]: Delete the comment.

File: src/components/AddBookForm.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
+import styled from 'styled-components/macro'
+
+import Button from './RectangleButton'
+import downArrow from '../assets/arrow-down.svg'
+
+export default function AddBookForm({ onCreateBook, onClickDownButton }) {
```
Comment (felix-horn) [inline src/components/AddBookForm.js]: I'd suggest to rename the `onCreateBook` function. We use the prefix `on*` for an eventlistener or s.th. that triggers another function. In this case, it is not the actual function per se, it is only used to pass the arguments to your `addBook` function, but it does not listen for an event. To make things easier, I'd just call it `addBook` as well and then in your App. js:
```
<AddBookForm
          addBook={addBook} 
...
```

File: src/components/AddBookForm.test.js [hunk 16]
```diff
@@ -0,0 +1,30 @@
+import AddBookForm from './AddBookForm'
+import user from '@testing-library/user-event'
+import { getByText, render } from '@testing-library/react'
+
+describe('AddBookForm', () => {
+  const onSubmitMock = jest.fn()
+
+  it('calls onSubmit with correct data and resets form', () => {
+    const { getByLabelText, getByText } = render(
+      <AddBookForm onCreateBook={onSubmitMock} />
```
Comment (felix-horn) [inline src/components/AddBookForm.test.js]: If you changed your addBook function as suggested avove, you'd also have to change it here:

      `<AddBookForm addBook={addBookMock} />`

and then of course rename the mock-function as well. (`addBookMock = jest.fn()`)

**Review by felix-horn — COMMENTED** [review]
Very nice work!!

**Pull request merged**